### PR TITLE
[RDY] Add nvim_set_keymap method to remote API

### PR DESCRIPTION
**Pull request by Yilin-Yang**
Meant to address #9136.

This would be my first contribution, so I'm still working to find my feet. I think that I've figured out the core of what I need to do (accept mappings from the user, parse them, and then pass them to `int do_map()`), but I'm unsure of a lot of the particulars.

Right now, I'm trying to figure out:

- What should the function signature be? (Should the user just pass in strings containing the "components" of the mapping? Pros: more flexible, extensible, and "natural", Cons: more complicated string parsing + related weird corner cases)
- How much input validation should we perform on the arguments the user provides? (We could infer that `:nnozfdxfdwse` is an attempt to `:nnoremap`, but should we still reject that as an invalid command?)
- How verbosely/granularly should we report errors to the user? (Okay to just return -1 no matter the error, or is it better to return, e.g. -2 for malformed arguments, etc.)

I'm also wondering if there are existing helper functions (beyond `get_map_mode`) on which I can offload much of the string parsing, since I'm not too satisfied with what I have right now. I'd appreciate any input or suggestions!

### CODE REVIEW

File: src/nvim/api/vim.c
Comment (mhinz) [inline src/nvim/api/vim.c]: No one-liners, please. :)

File: src/nvim/api/vim.c
Comment (mhinz) [inline src/nvim/api/vim.c]:
```suggestion
  bool is_noremap = STRNCMP(p + 1, "noremap", 7) == 0;
```

File: src/nvim/api/vim.c
Comment (mhinz) [inline src/nvim/api/vim.c]:
```suggestion
  bool is_unmap = map_cmd.size && p[0] == 'u';
```

File: src/nvim/tui/tui.c
Comment (mhinz) [inline src/nvim/tui/tui.c]: `tcsetpgrp()` is not available on Windows.

File: src/nvim/api/vim.c
Comment (mhinz) [inline src/nvim/api/vim.c]: You could add `Error *err` to the list of arguments. Then use `api_set_error()` later.

**Comment by bfredl** [comment]
Very nice! We probably want an `options` dictionary for less fundamental args like `silent` or `nowait`. A map of string keys is easier to deal with than string concatenation and parsing. Likewise we would want to split up `map_cmd` into its components, `mode` and separate `nore` flag. Unmap should likely be a separate API function. Possibly, `iabbr` and `cabbr` could be regarded as `mode` values.

File: src/nvim/api/vim.c
Comment (Yilin-Yang) [reply src/nvim/api/vim.c]: That's fair, haha! No worries, I'll fix that before I RFC!

File: src/nvim/api/vim.c
Comment (bfredl) [inline src/nvim/api/vim.c]: We would like a separate `nvim_buf_set_keymap` for this. Compare how `nvim_get_keymap` and `nvim_buf_get_keymap` is implemented. Most of the docs can stay at `nvim_set_keymap`, and the `nvim_buf` variant can just reference it.

File: src/nvim/api/vim.c
Comment (bfredl) [inline src/nvim/api/vim.c]: typically `opts` is the last arg (but before `Error`). Urrecognized keys should be an error.

File: src/nvim/api/vim.c
Comment (bfredl) [inline src/nvim/api/vim.c]: CI will give an error if this is incorrect, don't worry :)

File: src/nvim/api/vim.c
Comment (Yilin-Yang) [reply src/nvim/api/vim.c]: While I would love to implement this, I don't think that I have the time to do so, at least for now. Would it be possible to split this off into a different issue/PR?

**Comment by Yilin-Yang** [comment]
I think that I'm close to done; aside from needing to update `doc/api.txt` (and rebasing + force-pushing, etc.), I feel satisfied with what we have right now. Are there any other potential changes that stand out?

Also, thanks so much for your help, @mhinz and @bfredl! Receiving so much useful feedback so quickly came as a very pleasant surprise, and it made working on this PR much less difficult than it might have been otherwise.

File: gdb_start.sh
Comment (bfredl) [inline gdb_start.sh]: don't include this file in the PR. (can use `git rm --cached` to remove it from the branch while keeping the physical file).

File: src/nvim/api/vim.c
Comment (bfredl) [inline src/nvim/api/vim.c]: `rhs` and `lhs` should ideally allow whitespace, this is really just a parsing issue of `do_map`.  A strategy could be to split `do_map` after the parsing, probaly after the `if (maptype == 1 !haskey) {` condition. Then this API function could skip the whitespace parsing entirely, and send in lhs/rhs with whitespace in it (which then means the same as `<space>`) . Though maybe `\n` should still be forbidden, it litteraly is the same as `<c-j>` but many users would probably expect it to be mean `<cr>`

File: src/nvim/api/vim.c
Comment (bfredl) [reply src/nvim/api/vim.c]: New API functions should ideally not hardcode `curbuf` usage. I don't think this should be so difficult to achieve. Make `do_map` (or the second half, if the split suggestion is followed) take `buf_T *buf` arg and use it everywhere instead of `curbuf` inside of the function.

File: src/nvim/api/buffer.c
Comment (bfredl) [inline src/nvim/api/buffer.c]: We don't need the 'buffer' key at all. if `nvim_buf_set_keymap` is used, the mapping is buffer-local, and with `nvim_set_keymap` the mapping is global.

File: src/nvim/api/vim.c
Comment (bfredl) [inline src/nvim/api/vim.c]: Pass -1 as first arg to signal global behavior, just like `nvim_get_commands` (User code can't pass in -1).

File: src/nvim/api/buffer.c
Comment (bfredl) [inline src/nvim/api/buffer.c]: This was a restriction of `:map` ex command parsing, where empty argument is impossible to type. For API function there is no reason to disallow empty rhs.

File: src/nvim/ascii.h
Comment (justinmk) [inline src/nvim/ascii.h]: does not seem to be used

**Comment by justinmk** [comment]
@Yilin-Yang This is really nicely done, thank you!

> * How much input validation should we perform on the arguments the user provides? (We could infer that `:nnozfdxfdwse` is an attempt to `:nnoremap`, but should we still reject that as an invalid command?)

The API typically is "rigid", avoid "smart" behavior. "Smart" (unpredictable) behavior is already covered by VimL, and we don't need more of that in the API.

> * How verbosely/granularly should we report errors to the user? (Okay to just return -1 no matter the error, or is it better to return, e.g. -2 for malformed arguments, etc.)

The convention for API functions is to call `api_set_error` on the `Error` parameter, this  manifests as an exception in all clients. I see that `nvim_set_keymap` returns an error code, but I would suggest `void` return type (if there is nothing else meaningful to return). Rather the `Error` message text should reflect granular error case. In #6150 we decided that API error messages will not be translated. (We may also introduce new error "types" in the future, but the error messages will be reliable in any case.)

File: src/nvim/getchar.h
Comment (justinmk) [inline src/nvim/getchar.h]: /// @see mapblock_T

File: src/nvim/api/buffer.c
Comment (justinmk) [inline src/nvim/api/buffer.c]: "shortname is too long: %s"

("given" is implied)

File: src/nvim/api/buffer.c
Comment (justinmk) [inline src/nvim/api/buffer.c]: "Invalid mode shortname: %s";

We try to use the same language everywhere. Instead of "unrecognized", we usually say "invalid".

File: src/nvim/api/buffer.c
Comment (justinmk) [inline src/nvim/api/buffer.c]: "Invalid (empty) LHS"

we don't use "!" in messages :)

File: test/functional/api/buffer_spec.lua
Comment (justinmk) [inline test/functional/api/buffer_spec.lua]: if it's not too much trouble could you move this to `api/keymap_spec.lua`? That's where the tests for  `nvim_[buf_]_get_keymap` live.

File: test/functional/api/keymap_spec.lua
Comment (justinmk) [inline test/functional/api/keymap_spec.lua]: `helpers.wait()` is a more deterministic and less-expensive way to do this.

File: src/nvim/api/private/helpers.c
Comment (justinmk) [inline src/nvim/api/private/helpers.c]:
```suggestion
      err_msg = "Invalid key: %s";
```

Error messages do not restate the method name.

**Review by justinmk — APPROVED** [review]

**Comment by bfredl** [comment]
I think API here is starting to get really good, but I have an issue about the naming. `nvim_get_keymap` returns _all_ mappings in a mode, so it seems "a keymap" refers to a set of mappings (`help mbyte-keymap` also uses the word in the same way). Should the function be called `nvim_add_mapping` and `nvim_del_mapping` ? 

At least bundling `unmap` and `noremap` in the same arg as is done now seems a bit weird. `nore` is essentially a modifier (like `<expr>`)  that became part of the command name because reasons, while `unmap` is like a separate command (it doesn't take the same args and options etc).

**Comment by justinmk** [comment]
> Should the function be called `nvim_add_mapping` and `nvim_del_mapping` ?

I think `keymap` is fine. It was chosen for `nvim_set_keymap` to avoid confusion with a dictionary "map". I would rather have get/set "duals" than  `get_keymap` and `set_mapping`, it is less discoverable.

**Comment by bfredl** [comment]
Ok, but I would stil opt for `nvim_set_keymap` and `nvim_del_keymap` then, it is essentially analogous to `set_var` and `del_var`.

**Comment by justinmk** [comment]
Will wait for `nvim_del_keymap` and then this is ready to merge.

**Comment by Yilin-Yang** [comment]
I went ahead and added `nvim_[buf_]del_keymap` functions. Since that obviates the need for `set_keymap` to accept `maptype = "u"`, `maptype` would then only be useful for setting `noremap`, so I went ahead with @bredl's suggestion to make `noremap` an opt and then got rid of `maptype` entirely. That was a good suggestion; originally I was just following the signature of `do_map`, but breaking from that does make for a cleaner interface.

> regarding the new MapArgument type: did you consider adding fields to mapblock_T instead?

Also a good suggestion that I didn't think of while coding 😅. I do think it's better to have a separate type, though, for clarity's sake: since `buf_do_map_explicit` (and the original `do_map`) sometimes allocates/inserts a new `mapblock_T` into the global hashtables and sometimes doesn't (if it can reuse an existing one), it might be unclear who's left with "ownership" of the struct if it were passed into `buf_do_map_explicit` as a parameter.


Thanks for your guys' help! (and also don't be afraid to ping me if there are issues in the code i just pushed!)

**Comment by justinmk** [comment]
Merged. Thank you @Yilin-Yang , this is really appreciated. 

I squashed into 1 commit because most of the intermediate commits were not meaningful. For big/complex PRs, it's good to do it as you did: small commits without force-push. Then before merge it might make sense to rebase into a few meaningful commits--but for this PR I didn't think that would be worth your time. I preserved  commit messages where relevant.

I will make a followup PR with a few minor nitpicks.

**Pull request closed without merging**
